### DISCUSSION TITLE
blog/react-query-the-bad-parts

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#114] [Announcements]
# blog/react-query-the-bad-parts

In this talk, maintainer Dominik will explore the other side—the less favorable aspects of React Query and situations where it may not be the best fit. No library is perfect; every choice involves trade-offs. By the end of this talk, you'll have a better understanding of React Query's limitations and why it remains a compelling choice despite them.

https://tkdodo.eu/blog/react-query-the-bad-parts

**Comment by MirasMustimov** [comment]
Tanstack query is so fine, I am enjoying data fetching a lot now. Before it used to be the least favorite part of FE. Thank you Tanstack team!

**Comment by sanprabet** [comment]
Bruh, ik is the best, but you did us wrong with that clickbait lol

**Comment by mateogalic112** [comment]
Thank you TkDodo for all the efforts with React-Query. I have been using it for 4 years straight, shipped dosens of large scale apps and enjoyed every second of it! Keep up with great work, best wishes from Croatia! 🙌

**Comment by ver1fy** [comment]
For the most part I've gotten along just fine with the declarative approach. However, I recently ran into a problem where neither a declarative or imperative approach seemed to work elegantly. Assume we have a grid component similar to [TanStack Table](https://tanstack.com/table/latest) or [AG Grid](https://www.ag-grid.com/). In each cell we want an `onClick` context menu which renders a list of available filters for the column. The column filters are returned by a server.

How do we pass the data to each context menu once it is opened? One option is to hoist some active column state up into the grid and use it to trigger a [dependant query](https://tanstack.com/query/latest/docs/framework/react/guides/dependent-queries). Problem is, this will cause unnecessary re-renders to the entire grid, especially if we add more complex state such as search to the context menu.

```
const CustomGrid = () => {
    const [filterMenuColumn, setFilterMenuColumn] = useState();
    const { data: filterOptions, isLoading: filterOptionsLoading } = useQuery({
        queryKey: ['filterOptions', filterMenuColumn],
        queryFn: getFilterOptions,
        enabled: !!filterMenuColumn,
    });

    return (
        <Grid
            onFilterMenuOpen={filterMenuColumn}
            filterOptionsLoading={filterOptionsLoading}
            filterOptions={filterOptions} />
    );
};
```

Alternatively, we could take an imperative approach:
```
const CustomGrid = () => {
    const fetchFilterItems = async (column) => {
        const filterItems = await queryClient.ensureQueryData({
            queryKey: ['filterOptions', column],
            queryFn: getFilterOptions,
        });
        return filterItems;
    };

    return (
        <Grid filterItemsGetter={fetchFilterItems} />
    );
};
```
Whilst we avoid hoisting state, we lose the ability to observe a query and there's no easy way to pass a loading state to the context menu. Libraries such as AG Grid rely on a [`filterValueGetter`](https://www.ag-grid.com/javascript-data-grid/filter-set-filter-list/#supplying-filter-values) callback to pass the values. Is anyone able to offer a different perspective?

**Reply by TkDodo** [reply]
Is a custom component for the `MenuItem` only possible so that you can do the query there? The Grid API of `filterOptions` and `filterOptionsLoading` doesn’t look very ergonomic.

**Reply by ver1fy** [reply]
My implementation assumes that the context menu would live inside `Grid`. Having said that, perhaps the `Grid` and context menu could be siblings like so:
```
const CustomGrid = () => {
    const [filterMenuAnchorEl, setFilterMenuAnchorEl] = useState(undefined);
    const [filterItems, setFilterItems] = useState();

    return (
        <>
            <Grid
                onFilterMenuOpen={setFilterMenuAnchorEl}
                filterItems={filterItems} />
            <GridFilterMenu
                anchorEl={filterMenuAnchorEl}
                onFilterItemsChange={setFilterItems} />
        </>
    );
};
```
The `useQuery` hook could live inside `<GridFilterMenu />`. One thing I don't like is that the general purpose `<Grid />` component is no longer self contained; each custom grid has to remember to include `<GridFilterMenu />`. Also, it might be necessary to have multiple implementations of `<GridFilterMenu />` if each custom grid has different query dependencies.

**Reply by TkDodo** [reply]
feels like you want compound components:

```
<Grid>
  <GridFilterMenu />
  <GridFilterMenu />
</Grid>
```

**Reply by ver1fy** [reply]
I'm not sure if multiple `<GridFilterMenu />` components would offer a solution. The [AG Grid example](https://www.ag-grid.com/example/) might give a better idea of the desired end result. If you click on the funnel icon within a column, you'll see a list of filter options appear. Those options we'll want to fetch and cache with React Query.

I imagine there only needs to be one filter menu component which receives the column via a prop. The difficulty is how and where to do the data fetching since we don't want to couple our grid and menu components to a specific query. In other words, the grid and filter menu shouldn't be opinionated about how we fetch the options. Each implementation of a grid (see `<CustomGrid />` above) should be given freedom over how it fetches filter options.

**Reply by TkDodo** [reply]
you’ve basically found your options already: you can either do it in the parent of the `Grid`, which might make the whole Grid re-render when new data for a filter comes in. You’d likely need to memoize the Grid component then.

Or, you can move data fetching into the component that renders the filters. That’s what I would do. You can also pass `queryOptions` to your filters if you don’t want to couple them, instructing them what to fetch when they are rendered.

**Comment by mrlubos** [comment]
Did I just get Dominikrolled?

**Comment by hossein-jorfi** [comment]
hello, somthing out of context. this github comment section is a html document. you can style it to match the website theme😄

**Reply by TkDodo** [reply]
sure, please file a PR 🙏 

https://github.com/tkdodo/blog
